### PR TITLE
Attempt to unify perf kernel and AOTriton's kernel.

### PR DESCRIPTION
**Pull request by xinyazhang**
## Major changes

* Introduce a new naming convention: `Constant_in_jit_but_variable_in_aot`
    - Please use `constexpr_or_i32`/`constexpr_or_f32`/`constexpr_or_bool` annotation for such types
* Add `mstore2d` to handle OOB writes of encoded softmax
* Use `Max_seqlen_k` in calculating the Philox offsets, avoiding overlapped offsets for Varlen.
* Replace `PERSISTENT` and `PERSISTENT_DYNAMIC` with `PERSISTENT_TYPE = 0/1/2`  for No, Fixed and Dynamic options
* Add `Num_seqlens` to support padded varlen (Still Rank-4 Tensor, but only part of sequences are used)
    - `Num_seqlens == 0`: Classical SDPA
    - `Num_seqlens > 0`: Conventional Varlen
    - `Num_seqlens < 0`: Padded Varlen
* Replace `IS_CAUSAL` with `CAUSAL_TYPE`:
    - `CAUSAL_TYPE == 0`: No causal
    - `CAUSAL_TYPE == 1`: Top-left alignment (PyTorch ME backend's default)
    - `CAUSAL_TYPE == 2`: Bottom-right alignment (main_perf's previous settings)
* Fix Nan when sm_scale=0 on certain bias pattern (See: https://github.com/ROCm/aotriton/issues/47)
* Rearrange order of arguments. Please only add one argument for each line, with two exceptions:
    - strides of the same tensor should be put in a single line. 
    - Basic SPDA arguments and all tensors can be compactly packed together.
* Add hipGraph support to PRNG
* Add type annotation `tl.uint64` to all strides except for alibi, in order to support large tensor.

## Unit Test Changes

* Change the Dimension 0 of bias tensor from 1 to actual batch size through `torch.expand`.
* Add test for bias tensor with real batches, i.e., allocate actual memory and fill different values in different batches
    - This is required by LLAMA 3.2
* Remove the assertion of bias tensor size
    - `tl.uint64` type annotation on strides is used to overcome this limit

### CODE REVIEW

**Comment by xinyazhang** [comment]
Note: the performance downgrades to 300TFLOPS after these changes that seemingly cosmetic
Will dig the reason.

**Comment by xinyazhang** [comment]
Confirmed the hipGraph support causes huge performance degeneration. Even if the code path is guarded with ENABLE_DROPOUT

**Comment by xinyazhang** [comment]
Changed the UT in https://github.com/ROCm/triton/pull/716/commits/a4e68374e5e01a2e0ef196b60158fcd3943b3dbc

This is required by LLAMA 3.2. which has different values in different batches of the bias tensor.
Ignoring batch index causes SWDEV-508774

Changes in related projects:
* https://github.com/ROCm/aotriton/pull/68
* https://github.com/pytorch/pytorch/pull/145508

**Comment by xinyazhang** [comment]
@jayfurmanek @vgokhale All UTs passed

**Comment by vgokhale** [comment]
Can you post performance before and after these changes for both the model and default cmd line args?

**Comment by xinyazhang** [comment]
`og-flash-attention.py` is the original FA file I checked out to make the comparison easier.

```
(py_3.10) xinyazha@1949bd361800:~/triton-rocm/python/perf-kernels$ sha256sum flash-attention.py og-flash-attention.py 
4754219a53db54c2eea70fc5cb94bf5fce393c286aa730e862734e2ad33ee240  flash-attention.py
244718ba8927a15aa11ea14f5b34af0c1fb0a923a8965a87c0650092bed37397  og-flash-attention.py
(py_3.10) xinyazha@1949bd361800:~/triton-rocm/python/perf-kernels$ python flash-attention.py 
fused-attention-fwd-d128-layoutbhsd:
    BATCH    HQ    HK  N_CTX_Q  N_CTX_K      triton       torch
0    16.0  16.0  16.0   1024.0   1024.0  290.451385  209.159602
1     8.0  16.0  16.0   2048.0   2048.0  348.721926  262.840479
2     4.0  16.0  16.0   4096.0   4096.0  370.112906  285.095079
3     2.0  16.0  16.0   8192.0   8192.0  381.123075  293.967613
4     1.0  16.0  16.0  16384.0  16384.0  378.565994  281.094768
5     2.0  48.0  48.0   1024.0   1024.0  253.586024  199.107892
6     2.0  48.0  48.0   2048.0   1024.0  308.630244  228.837054
7     2.0  48.0  48.0   4096.0   8192.0  362.719665  286.586611
8     2.0  48.0  48.0   8192.0   4096.0  378.872578  301.687999
9     2.0  48.0  48.0  16384.0   8192.0  390.812812  316.858139
10    8.0  16.0  16.0   1989.0  15344.0  290.809432  262.368374
11    4.0  16.0  16.0   4097.0    163.0  133.408037   91.389727
12    2.0  16.0  16.0   8122.0   2159.0  275.348037  264.067166
13    1.0  16.0  16.0  16281.0      7.0    6.050914    6.945960
14    2.0  48.0  48.0   1021.0   1020.0  210.570050  192.480003
15    2.0  48.0  48.0   2001.0   2048.0  323.025914  254.007358
16    2.0  48.0  48.0   3996.0   9639.0  292.943909  274.532452
17    2.0  48.0  48.0   8181.0   1021.0  256.953097  259.333202
(py_3.10) xinyazha@1949bd361800:~/triton-rocm/python/perf-kernels$ python og-flash-attention.py 
fused-attention-fwd-d128-layoutbhsd:
    BATCH    HQ    HK  N_CTX_Q  N_CTX_K      triton       torch
0    16.0  16.0  16.0   1024.0   1024.0  289.643802  209.040048
1     8.0  16.0  16.0   2048.0   2048.0  348.803770  262.284504
2     4.0  16.0  16.0   4096.0   4096.0  371.181808  284.123233
3     2.0  16.0  16.0   8192.0   8192.0  381.649394  294.678136
4     1.0  16.0  16.0  16384.0  16384.0  380.010986  281.142287
5     2.0  48.0  48.0   1024.0   1024.0  252.326097  198.872377
6     2.0  48.0  48.0   2048.0   1024.0  306.027914  228.772620
7     2.0  48.0  48.0   4096.0   8192.0  364.630607  286.723525
8     2.0  48.0  48.0   8192.0   4096.0  379.244673  301.998962
9     2.0  48.0  48.0  16384.0   8192.0  392.781918  316.020913
10    8.0  16.0  16.0   1989.0  15344.0  281.440980  261.661028
11    4.0  16.0  16.0   4097.0    163.0  136.364422   91.629028
12    2.0  16.0  16.0   8122.0   2159.0  264.899686  264.636067
13    1.0  16.0  16.0  16281.0      7.0    6.896461    6.926344
14    2.0  48.0  48.0   1021.0   1020.0  210.527696  191.724849
15    2.0  48.0  48.0   2001.0   2048.0  322.386144  254.093665
16    2.0  48.0  48.0   3996.0   9639.0  282.014099  275.197678
17    2.0  48.0  48.0   8181.0   1021.0  256.838505  260.254701
(py_3.10) xinyazha@1949bd361800:~/triton-rocm/python/perf-kernels$ python flash-attention.py -model all
fused-attention-fwd-layoutbhsd:
         model  BATCH   HQ  HK  N_CTX_Q  N_CTX_K  D_HEAD      triton       torch
0    llama3-8B      1   32   8     4096     4096     128  363.091576  285.330544
1   llama3-70B      1   64   8     4096     4096     128  385.504352  285.391074
2  llama3-405B      1  128   8     4096     4096     128  392.309230  306.246575
(py_3.10) xinyazha@1949bd361800:~/triton-rocm/python/perf-kernels$ python og-flash-attention.py -model all
fused-attention-fwd-layoutbhsd:
         model  BATCH   HQ  HK  N_CTX_Q  N_CTX_K  D_HEAD      triton       torch
0    llama3-8B      1   32   8     4096     4096     128  361.328059  284.975282
1   llama3-70B      1   64   8     4096     4096     128  385.783594  285.467320
2  llama3-405B      1  128   8     4096     4096     128  390.250257  305.214373
```
@vgokhale here it is

File: python/perf-kernels/flash-attention.py [hunk 1]
```diff
@@ -425,31 +554,99 @@ def get_autotune_configs():
...
     key=autotune_keys,
     use_cuda_graph=True,
 )
 @triton.jit
-def attn_fwd(Q, K, V, bias, SM_SCALE: tl.constexpr, L, Out, stride_qz, stride_qh, stride_qm, stride_qk, stride_kz,
-             stride_kh, stride_kn, stride_kk, stride_vz, stride_vh, stride_vk, stride_vn, stride_oz, stride_oh,
-             stride_om, stride_on, stride_bz, stride_bh, stride_bm, stride_bn, stride_az, stride_ah, Q_descale,
-             K_descale, P_scale, P_descale, V_descale, cu_seqlens_q, cu_seqlens_k, dropout_p, philox_seed,
-             PERSISTENT: tl.constexpr, PERSISTENT_DYNAMIC: tl.constexpr, atomic_counter, NUM_CU: tl.constexpr,
-             GRID_CU_MULTIP: tl.constexpr, B: tl.constexpr, philox_offset_base, encoded_softmax, alibi_slopes,
```
Comment (juuso-oskari) [inline python/perf-kernels/flash-attention.py:439]: B shouldnt be a tl.constexpr? Because it will end up compiling for each new batch size

**Comment by Chi-Chu319** [comment]
@xinyazhang Could you also benchmark with -causal? I think that's where you get the most perf boost from

File: python/perf-kernels/flash-attention.py (hunk 1 above)
Comment (juuso-oskari) [reply python/perf-kernels/flash-attention.py:439]: Can we delay this PR until I fix the persistent approach in the MLA decode? I found an issue there, and solving it will benefit the persistent approach here aswell.
(details: the persistent approach is currently increasing register usage by a large amount needlessly. I checked that its also happening with the FA persistent and will result in autotuning choosing smaller and non-optimal block sizes to deal with the register usage.)

**Comment by xinyazhang** [comment]
@juuso-oskari @Chi-Chu319 the work related to this PR is currently paused due to on-going fixes for https://github.com/ROCm/triton-internal/issues/560#issuecomment-2654030077
I'll resume work once the performance problem for seqlen=1024 was fixed.
